linalg/hypre_parcsr.cpp

```diff
@@ -1520,10 +1520,14 @@ hypre_ParCSRMatrixSum(hypre_ParCSRMatrix *A,
    hypre_CSRMatrix *A_offd = hypre_ParCSRMatrixOffd(A);
    hypre_CSRMatrix *B_diag = hypre_ParCSRMatrixDiag(B);
    hypre_CSRMatrix *B_offd = hypre_ParCSRMatrixOffd(B);
+   HYPRE_Int ncols_B_offd  = hypre_CSRMatrixNumCols(B_offd);
    HYPRE_Int error;
 
    error = hypre_CSRMatrixSum(A_diag, beta, B_diag);
-   error = error ? error : hypre_CSRMatrixSum(A_offd, beta, B_offd);
+   if (ncols_B_offd > 0) /* treat B_offd as zero if it has no columns */
+   {
+      error = error ? error : hypre_CSRMatrixSum(A_offd, beta, B_offd);
+   }
```

Comment:
This fix seems to be masking some other issue: if `hypre_CSRMatrixSum` returns an error then `A_offd` and `B_offd` have different sizes which should not happen.

Note that `hypre_ParCSRMatrixSum` and its wrapper `HypreParMatrix::Add` (as stated in their doxygen documentation) assume that `A` and `B` use the same column and row partitionings **and** have the same `col_map_offd`. It looks like the second part (about `col_map_offd`) is what is not true, so technically this method should not be used. The right function to use in such cases is `ParAdd` which uses `hypre_ParCSRMatrixAdd` which in turn supports the case when `col_map_offd` are different.

Reply:
Actually, it looks like the doxygen comments for `ParAdd` and `HypreParMatrix *Add(double alpha, const HypreParMatrix &A, double beta,  const HypreParMatrix &B)` have not been updated to reflect the fact that `hypre_ParCSRMatrixAdd` now supports the case of different `col_map_offd` arrays.

Reply:
Hmm, actually, `HypreParMatrix *Add(double alpha, const HypreParMatrix &A, double beta, const HypreParMatrix &B)` uses `hypre_ParCSRMatrixSum` as well, so different `col_map_offd` will not work.

Reply:
I also thought that `A_offd` and `B_offd` should not have different sizes.  I'm still a novice with DG methods but I believe it is correct that a DG mass matrix has no `offd` entries but a DG diffusion matrix obvious will.  I ran into this problem when trying to form `M + dt S`.

Reply:
Yes, since `M` and `S` have different `col_map_offd` (`M`'s offd matrix is empty so its `col_map_offd` is also empty), you can't use `HypreParMatrix::Add`. You should use `ParAdd` instead.

Reply:
I understand that but in this special case it's much more efficient to recognize that adding zero to an existing matrix requires no work at all so we should simply allow it rather than producing an error.

I would also argue that this is in accordance with how we handle the diagonal portion of HYPRE matrices in `hypre_CSRMatrixSum`.  Specifically, this function returns an error if it finds an entry in `B` that doesn't exist in `A` but not the other way around, why should it?  The result is that `B` is allowed to have a smaller sparsity pattern than `A` in the diagonal portion.  The change proposed in this PR is similar because it also allows the off-diagonal portion of `B` to have a smaller sparsity pattern than `A`.  Much smaller in this case but just as valid to ignore.

Reply:
OK, this makes sense.

It will be good to add this to the documentation of `HypreParMatrix::Add` and `hypre_ParCSRMatrixSum`: `A` and `B` must have the same `col_map_offd`, or `B` has empty off-diagonal.

Reply:
Good point.  I'll take care of that.  Thanks!